internal/event/dispatch_service.go

```diff
@@ -31,209 +30,91 @@ type Event interface {
 
 type Handler func(event Event)
 
-type controlType int
-
-const (
-	subscribe controlType = iota
-	subscribeSync
-	unsubscribe
-)
-
-const (
-	created int32 = iota
-	ready
-	stopped
-)
-
-type controlMessage struct {
-	handler        Handler
-	eventName      string
-	controlType    controlType
-	subscriptionID int64
+type subscription struct {
+	handler   Handler
+	orderChan chan Event
 }
 
 type DispatchService struct {
-	logger            logger.Logger
-	syncSubscriptions map[string]map[int64]Handler
-	eventCh           chan Event
-	controlCh         chan controlMessage
-	doneCh            chan struct{}
-	subscriptions     map[string]map[int64]Handler
-	state             int32
+	logger        logger.Logger
+	subscriptions map[string]map[int64]subscription
+	eventMu       *sync.RWMutex
+	isClosed      bool
 }
 
+// NewDispatchService creates a dispatch service with the following properties.
+//1- It is async for the caller of Publish. Meaning Publish will not be blocked.
+//It can be blocked because we don't have inifnite channels. The queue size is 1024.
+//2- Events fired from the same thread with the same event name will be handled in the same order.
+//One will finish, then other will start.
+//3 - If we block an event, it should not block all events only ones that are related to same EventName.
+//4 - A close after publish in the same thread waits for published item to be handled(finished) .
 func NewDispatchService(logger logger.Logger) *DispatchService {
 	service := &DispatchService{
-		subscriptions:     map[string]map[int64]Handler{},
-		syncSubscriptions: map[string]map[int64]Handler{},
-		eventCh:           make(chan Event, 1024),
-		controlCh:         make(chan controlMessage, 1024),
-		doneCh:            make(chan struct{}),
-		state:             created,
-		logger:            logger,
+		subscriptions: map[string]map[int64]subscription{},
+		eventMu:       &sync.RWMutex{},
+		logger:        logger,
 	}
-	startCh := make(chan struct{})
-	go service.start(startCh)
-	<-startCh
-	atomic.StoreInt32(&service.state, ready)
 	return service
 }
-
+// Stop the DispatchService and make sure that all the successful Publish events are handled before close
```

Comment:
No, this is not guranteed. Closing a channel does not wait for all events in it to be consumed. When the process terminates, the goroutines that read from `orderChan` are terminated as well. So, there may still be events waiting in `orderChan`.

Reply:
Yes, you are right. I assumed to go behaviour wrong. I have added waitgroup to wait for this. see https://github.com/hazelcast/hazelcast-go-client/pull/652/commits/3a9d317844176b198f99e65912b90f7d5161b35b